module/VuFind/src/VuFind/Db/Service/ChangeTrackerService.php

```diff
@@ -0,0 +1,286 @@
+     *
+     * @param string $core The Solr core holding the record.
+     * @param string $id   The ID of the record being indexed.
+     *
+     * @return ChangeTracker|false
+     */
+    public function retrieveOrCreate($core, $id)
+    {
+        $row = $this->retrieve($core, $id);
+        if (empty($row)) {
+            $now = new \DateTime('now', new \DateTimeZone('UTC'));
+            $row = $this->createEntity()
+                ->setCore($core)
+                ->setId($id)
+                ->setFirstIndexed($now)
+                ->setLastIndexed($now);
+            try {
+                $this->persistEntity($row);
+            } catch (\Exception $e) {
+                $this->logError('Could not save change tracker record: ' . $e->getMessage());
+                return false;
+            }
+        }
+        return $row;
+    }
+
+    /**
+     * Update the change tracker table to indicate that a record has been deleted.
+     *
+     * The method returns the updated/created row when complete.
+     *
+     * @param string $core The Solr core holding the record.
+     * @param string $id   The ID of the record being indexed.
+     *
+     * @return ChangeTracker|false
+     */
+    public function markDeleted($core, $id)
+    {
+        // Get a row matching the specified details:
+        $row = $this->retrieveOrCreate($core, $id);
+
+        // If the record is already deleted, we don't need to do anything!
+        if (!empty($row->getdeleted())) {
+            return $row;
+        }
+
+        // Save new value to the object:
+        $row->setDeleted(new \DateTime('now', new \DateTimeZone('UTC')));
+        try {
+            $this->persistEntity($row);
+        } catch (\Exception $e) {
+            $this->logError('Could not update the deleted time: ' . $e->getMessage());
+            return false;
+        }
+        return $row;
+    }
+
+    /**
+     * Update the change_tracker table to reflect that a record has been indexed.
+     * We need to know the date of the last change to the record (independent of
+     * its addition to the index) in order to tell the difference between a
+     * reindex of a previously-encountered record and a genuine change.
+     *
+     * The method returns the updated/created row when complete.
+     *
+     * @param string $core   The Solr core holding the record.
+     * @param string $id     The ID of the record being indexed.
+     * @param int    $change The timestamp of the last record change.
+     *
+     * @return ChangeTracker
+     */
+    public function index($core, $id, $change)
+    {
+        // Get a row matching the specified details:
+        $row = $this->retrieveOrCreate($core, $id);
+
+        // Flag to indicate whether we need to save the contents of $row:
+        $saveNeeded = false;
+        $utcTime = \DateTime::createFromFormat('U', $change, new \DateTimeZone('UTC'));
+        $now = new \DateTime('now', new \DateTimeZone('UTC'));
```

Comment:
$now is only used inside the if statement on line 217, so it's probably better to define it in that scope, a little later.